### ISSUE TITLE
Spring Boot & Freemarker 2.3.30

### ISSUE THREAD

**Issue by pgervaise** [spring-projects/spring-boot#22166]
Hi,

When using Spring Boot 2.3.1 with Freemarker (so the version of Freemarker is 2.3.30) I got this error in console log :

> 2020-06-30 10:23:43,243 **ERROR** [freemarker.beans] (default task-7) **BeansWrapper.incompatibleImprovements was set to the object returned by Configuration.getVersion()**. That defeats the purpose of incompatibleImprovements, and makes upgrading FreeMarker a potentially breaking change. Also, this probably won't be allowed starting from 2.4.0. Instead, set incompatibleImprovements to the highest concrete version that's known to be compatible with your application.

It's shown for every page, it's just a log but it's annoying AND it indicate that a futur version of Freemarker will not continue to work ... 

I tried to use Spring Boot 2.2.8 but it's the same thing. 

To resolve my issue I finally forced the freemarker version to be 2.3.29 by using this in pom.xml :

	<properties>
		<freemarker.version>2.3.29</freemarker.version>
	</properties>

It would be nice if the problem is corrected in a futur version.
Thank you.

**Comment by wilkinsona** [comment]
Thanks for the report. Unfortunately, I cannot reproduce the behaviour you have described. No errors are logged when running of Freemarker smoke test application, and debugging it shows that the incompatible improvements version is 2.3.0 while the configuration version is 2.3.30.  If you would like us to spend some more time investigating, please spend some time providing a complete yet minimal sample that reproduces the problem. You can share it with us pushing it to a separate repository on GitHub or by zipping it up and attaching it to this issue.

**Comment by pgervaise** [comment]
I reduced my enterprise project to the strict minimum (and without personal informations) and I obtained a mini-project where the problem occurs.

The code source is available here : [https://github.com/pgervaise/web-planning-test-freemarker](https://github.com/pgervaise/web-planning-test-freemarker)

When using spring boot 2.2.5 : no warning (because of Freemarker 2.3.29)
When using spring boot 2.2.8 or 2.3.1 : there is a warning (because of Freemarker 2.3.30)

I discovered that it's inherited by the class *FreemarkerConfig* when configuring Freemarker for Spring Boot. The warning is generated by the class *PlanningFreeMarkerView* at line 29.

I guess it should have a trick to do the same thing in the right way (for freemarker 2.3.30+).

**Comment by wilkinsona** [comment]
Thanks for the sample. The behaviour that you are seeing is due to [this change in FreeMarker 2.3.30](https://github.com/apache/freemarker/commit/e816f81e55d9cd00b7c17525fd665eff1f630bb2) and how you're [creating the BeansWrapperBuilder](https://github.com/pgervaise/web-planning-test-freemarker/blob/c039b054e2f25fc3a663ef99fd08d9a4f0e6f5f5/src/main/java/fr/penelope/planning/web/config/freemarker/PlanningFreeMarkerView.java#L29):

```java
BeansWrapper wrapper = new BeansWrapperBuilder(Configuration.getVersion()).build();
```

As recommended in FreeMarker's javadoc, you should pass in a fixed version rather than what's returned from `Configuration.getVersion()`.

**Comment by pgervaise** [comment]
Thank you for your help.

I changed `Configuration.getVersion()` by `new Version(Configuration.getVersion().toString())` and it works.

It's a little bit tricky but it let Spring MVC managing the version of freemarker used. Found it in [Spring Context Support / FreeMarkerConfigurationFactory class](https://github.com/spring-projects/spring-framework/blob/f13ab6d18c78e593d5c41cd6355b2e6710b27b01/spring-context-support/src/main/java/org/springframework/ui/freemarker/FreeMarkerConfigurationFactory.java#L327).
